File: app/src/main/java/com/mouredev/weeklychallenge2022/Challenge47.py
"""
package com.mouredev.weeklychallenge2022

/*
 * Reto #47
 * VOCAL MÁS COMÚN
 * Fecha publicación enunciado: 21/11/22
 * Fecha publicación resolución: 28/11/22
 * Dificultad: FÁCIL
 *
 * Enunciado: Crea un función que reciba un texto y retorne la vocal que más veces se repita.
 * Si no hay vocales podrá devolver vacío.
 *
 * Información adicional:
 * - Usa el canal de nuestro Discord (https://mouredev.com/discord) "🔁reto-semanal"
 *   para preguntas, dudas o prestar ayuda a la comunidad.
 * - Tienes toda la información sobre los retos semanales en
 *   https://retosdeprogramacion.com/semanales2022.
 *
 */
"""

def normalizatexto(texto):
    texto = texto.lower()
    texto = texto.replace("á","a")
    texto = texto.replace("é","e")
    texto = texto.replace("í","i")
    texto = texto.replace("ó","o")
    texto = texto.replace("ú","u")
    return texto


def vocalmasrepetida(dict):
    maximo = max(dict, key=dict.get)
    if dict[maximo]==0:
        return None
    else:
        return maximo


def cuentavocal(texto):
    dict = {"a":0,"e":0,"i":0,"o":0,"u":0}
    lista = list(texto)

    for i in lista:
        vocales = ("a","e","i","o","u")
        if i in vocales:
            dict[i]+=1
    return dict


if __name__=="__main__":
    texto = input("Ingresa una oración: ")
    texto = normalizatexto(texto)
    totalvocales = cuentavocal(texto)
    max = vocalmasrepetida(totalvocales)
    if max==None:
        print ("No hay vocales en la oración")
    else:
        print (f'La vocal más repetida es la: "{max}" y se repite {totalvocales[max]} veces en la oración')

File: app/src/main/java/com/mouredev/weeklychallenge2022/Challenge47.rb
# frozen_string_literal: true

#
# Reto #47
# VOCAL MÁS COMÚN
# Fecha publicación enunciado: 21/11/22
# Fecha publicación resolución: 28/11/22
# Dificultad: FÁCIL
#
# Enunciado: Crea un función que reciba un texto y retorne la vocal que más veces se repita.
# Si no hay vocales podrá devolver vacío.
#
# Información adicional:
# - Usa el canal de nuestro Discord (https://mouredev.com/discord) "🔁reto-semanal"
#   para preguntas, dudas o prestar ayuda a la comunidad.
# - Tienes toda la información sobre los retos semanales en
#   https://retosdeprogramacion.com/semanales2022.
#
#

# class Vowel
class Vowel
  def most_repeated_vowels(input)
    max_vowel(count_vowels(input))
  end

  def count_vowels(input)
    vowels = /[aeiou]/
    # vowels_numbers = {}
    # input.downcase.tr('áéíóúü', 'aeiouu').each_char do |elem|
    # input.downcase.gsub(/[áéíóúü]/, 'á' => 'a', 'é' => 'e', 'í' => 'i', 'ó' => 'o', 'ú' => 'u',
    # 'ü' => 'u').scan(vowels).each do |letter|
    # input.downcase.gsub(/[áéíóúü]/, 'á' => 'a', 'é' => 'e', 'í' => 'i', 'ó' => 'o', 'ú' => 'u',
    #                                 'ü' => 'u').each_char do |letter|
    # next unless letter.match(vowels)

    #   vowels_numbers[letter] = 0 unless vowels_numbers.keys.include?(letter)
    #   vowels_numbers[letter] += 1
    # end

    input.downcase.gsub(/[áéíóúü]/, 'á' => 'a', 'é' => 'e', 'í' => 'i', 'ó' => 'o', 'ú' => 'u',
                                    'ü' => 'u').scan(vowels).each.with_object(Hash.new(0)) do |letter, sum|
      sum[letter] += 1
    end
  end

  def max_vowel(counted_vowels)
    # max_value = 0
    # max_vowel = ''
    # counted_vowels.each do |key, value|
    #   if value > max_value
    #     max_value = value
    #     max_vowel = key
    #   end
    # end
    # max_vowel
    return '' if counted_vowels.empty?

    counted_vowels.max { |a, b| a[1] <=> b[1] }[0]
  end
end
